src/cert.rs

```diff
@@ -163,51 +166,61 @@ enum Understood {
 }
 
 fn remember_extension<'a>(
-    cert: &mut Cert<'a>, extn_id: untrusted::Input, value: untrusted::Input<'a>,
+    cert: &mut Cert<'a>, extn_id: untrusted::Input<'a>, value: untrusted::Input<'a>,
 ) -> Result<Understood, Error> {
     // We don't do anything with certificate policies so we can safely ignore
     // all policy-related stuff. We assume that the policy-related extensions
     // are not marked critical.
 
-    // id-ce 2.5.29
-    static ID_CE: [u8; 2] = oid![2, 5, 29];
+    let mut ret = Ok(Understood::Yes);
 
-    if extn_id.len() != ID_CE.len() + 1 || !extn_id.as_slice_less_safe().starts_with(&ID_CE) {
-        return Ok(Understood::No);
-    }
+    let extension_id = &*extn_id.as_slice_less_safe();
 
-    let out = match *extn_id.as_slice_less_safe().last().unwrap() {
+    let out = match extension_id {
         // id-ce-keyUsage 2.5.29.15. We ignore the KeyUsage extension. For CA
         // certificates, BasicConstraints.cA makes KeyUsage redundant. Firefox
         // and other common browsers do not check KeyUsage for end-entities,
         // though it would be kind of nice to ensure that a KeyUsage without
         // the keyEncipherment bit could not be used for RSA key exchange.
-        15 => {
-            return Ok(Understood::Yes);
+        [85, 29, 15] => {
+             return ret;
         },
 
         // id-ce-subjectAltName 2.5.29.17
-        17 => &mut cert.subject_alt_name,
+        [85, 29, 17] => &mut cert.subject_alt_name,
 
         // id-ce-basicConstraints 2.5.29.19
-        19 => &mut cert.basic_constraints,
+        [85, 29, 19] => &mut cert.basic_constraints,
 
         // id-ce-nameConstraints 2.5.29.30
-        30 => &mut cert.name_constraints,
+        [85, 29, 30] => &mut cert.name_constraints,
 
         // id-ce-extKeyUsage 2.5.29.37
-        37 => &mut cert.eku,
+        [85, 29, 37] => &mut cert.eku,
 
+        //_ => return Ok(Understood::No),
         _ => {
-            return Ok(Understood::No);
+            ret = Ok(Understood::No);
+            let value = value.read_all(Error::BadDER, |value| {
+                Ok(value.read_bytes_to_end())
+            })?;
+            match cert.unrecognized_extensions.insert(extension_id, value) {
+                Some(_) => {
+                    // the certificate contains more than one instance of this extension
+                    ret = Err(Error::ExtensionValueInvalid);
```

Comment:
Why not just

```rust
return Err(Error::ExtensionValueInvalid);
```

Reply:
My goal was to be consistent throughout the function. Instead of sometimes returning a previously set value, and sometimes returning a value directly, just always return a previously set value.

I think this may have been more relevant in a previous iteration of the function.